File: packages/java/src/main/java/com/tesote/sdk/v1/TransactionsClient.java
package com.tesote.sdk.v1;

import com.fasterxml.jackson.databind.JsonNode;
import com.tesote.sdk.Transport;
import com.tesote.sdk.internal.Json;
import com.tesote.sdk.internal.QueryParams;
import com.tesote.sdk.models.Transaction;
import com.tesote.sdk.models.TransactionsPage;

import java.time.Duration;
import java.util.Objects;

/**
 * v1 transactions: list-for-account (cursor) and read-by-id.
 */
public final class TransactionsClient {
    private static final Duration GET_CACHE = Duration.ofMinutes(5);

    private final Transport transport;

    public TransactionsClient(Transport transport) { this.transport = transport; }

    public TransactionsPage listForAccount(String accountId) {
        return listForAccount(accountId, new ListParams());
    }

    public TransactionsPage listForAccount(String accountId, ListParams params) {
        Objects.requireNonNull(accountId, "accountId");
        Objects.requireNonNull(params, "params");
        Transport.Options opts = Transport.Options.get(
                "/v1/accounts/" + AccountsClient.encode(accountId) + "/transactions")
                .query(params.toQuery());
        JsonNode node = transport.request(opts);
        return Json.treeToValue(node, TransactionsPage.class);
    }

    public Transaction get(String id) {
        Objects.requireNonNull(id, "id");
        Transport.Options opts = Transport.Options.get(
                "/v1/transactions/" + AccountsClient.encode(id))
                .cacheTtl(GET_CACHE);
        return Json.treeToValue(transport.request(opts), Transaction.class);
    }

    /** Cursor + date filter parameters for {@link #listForAccount(String, ListParams)}. */
    public static final class ListParams {
        public String startDate;
        public String endDate;
        public String scope;
        public Integer page;
        public Integer perPage;
        public String transactionsAfterId;
        public String transactionsBeforeId;

        public ListParams startDate(String v) { this.startDate = v; return this; }
        public ListParams endDate(String v) { this.endDate = v; return this; }
        public ListParams scope(String v) { this.scope = v; return this; }
        public ListParams page(int v) { this.page = v; return this; }
        public ListParams perPage(int v) { this.perPage = v; return this; }
        public ListParams transactionsAfterId(String v) { this.transactionsAfterId = v; return this; }
        public ListParams transactionsBeforeId(String v) { this.transactionsBeforeId = v; return this; }

        java.util.Map<String, String> toQuery() {
            return QueryParams.of()
                    .put("start_date", startDate)
                    .put("end_date", endDate)
                    .put("scope", scope)
                    .put("page", page)
                    .put("per_page", perPage)
                    .put("transactions_after_id", transactionsAfterId)
                    .put("transactions_before_id", transactionsBeforeId)
                    .build();
        }
    }
}

File: packages/csharp/src/V1/TransactionsClient.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Tesote.Sdk.Internal;
using Tesote.Sdk.Models;

namespace Tesote.Sdk.V1;

/// <summary>v1 Transactions resource: per-account listing and single-id lookup.</summary>
public sealed class TransactionsClient
{
    private const string BasePath = "/v1/transactions";
    private const string AccountsBasePath = "/v1/accounts";

    private readonly Transport _transport;

    internal TransactionsClient(Transport transport)
    {
        _transport = transport;
    }

    /// <summary>List transactions for a single account, with optional date and cursor filters.</summary>
    public async Task<TransactionListResponse> ListForAccountAsync(
        string accountId,
        string? startDate = null,
        string? endDate = null,
        string? scope = null,
        int? page = null,
        int? perPage = null,
        string? transactionsAfterId = null,
        string? transactionsBeforeId = null,
        CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(accountId);
        var query = new QueryBuilder()
            .Add("start_date", startDate)
            .Add("end_date", endDate)
            .Add("scope", scope)
            .Add("page", page)
            .Add("per_page", perPage)
            .Add("transactions_after_id", transactionsAfterId)
            .Add("transactions_before_id", transactionsBeforeId)
            .BuildOrNull();

        var opts = RequestOptions.Get(AccountsBasePath + "/" + Uri.EscapeDataString(accountId) + "/transactions");
        opts.Query = query;
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<TransactionListResponse>(node);
    }

    /// <summary>Fetch a single transaction by id.</summary>
    public async Task<Transaction> GetAsync(string transactionId, CancellationToken ct = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(transactionId);
        var opts = RequestOptions.Get(BasePath + "/" + Uri.EscapeDataString(transactionId));
        opts.CacheTtl = TimeSpan.FromMinutes(5);
        var node = await _transport.RequestAsync(opts, ct).ConfigureAwait(false);
        return Json.Deserialize<Transaction>(node);
    }
}
